### PR TITLE
Migrate normalization family to ProgramDescriptor framework

### LINKED ISSUE #42320: [Descriptor Migration] normalization/batch_norm
Migrate `normalization/batch_norm` from the legacy `OldDeviceOperation` pattern to `ProgramDescriptor`.

Part of #42203.

## Files
`ttnn/cpp/ttnn/operations/normalization/batch_norm/`

## Resources
- Detection: `python3 scripts/detect_legacy_device_op.py <file>`
- Docs: https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/adding_new_ttnn_operation.html

### PR DESCRIPTION
**Pull request by dgomezTT**
Closes #42320, #42321, #42322, #42323, #42324. Part of #42203.

Migrates the normalization family (`batch_norm`, `groupnorm`, `layernorm`, `layernorm_distributed`, `softmax`) to `ProgramDescriptor` and adopts the `BufferBinding` fast cache-hit path from #42992.

## Tests

486 passed / 0 failed across `test_softmax`, `test_softmax_interleaved`, `test_layernorm`, `test_layernorm_sharded`, `test_rmsnorm`, `test_group_norm`, `test_group_norm_DRAM`.

## Performance

Host-side dispatch only (device kernel time unchanged). Hot cache, n=200×3, median, Wormhole B0, shape `[1,1,64,64]` (or noted). Compared current branch ("descriptor + binding") vs `origin/main` ("legacy"):

| Op | Legacy main (μs) | Descriptor + binding (μs) | Δ |
|---|---:|---:|---:|
| `layer_norm` | 24.72 | 15.49 | **−37.3%** |
| `rms_norm` | 19.45 | 24.46 | +25.8% |
| `softmax dim=-1` | 24.26 | 34.42 | +41.9% |
| `softmax dim=-2` | 18.83 | 18.74 | −0.5% |
| `softmax dim=1 [1,4,64,64]` | 27.54 | 28.44 | +3.3% |
| `batch_norm (eval)` | 21.73 | 21.18 | −2.5% |
| `groupnorm` | _setup error_ | _setup error_ | — |
| `layer_norm_distributed pre/post` | _multi-device_ | _multi-device_ | — |

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:dgomez/migrate-normalization-to-descriptor)

### CODE REVIEW

**Review by bbradelTT — APPROVED** [review]
LTGM
Please make sure https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml and https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml are green or fail in the same way as main.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_program_factory_attention_optimized_sharded.cpp [hunk 1]
```diff
@@ -132,14 +133,13 @@ SoftmaxShardedProgramFactoryAttentionOptimized::cached_program_t SoftmaxShardedP
     uint32_t start_core_y = 0;
     uint32_t num_cores_c = program_config.compute_with_storage_grid_size.x;
     uint32_t num_cores_r = program_config.compute_with_storage_grid_size.y;
-    uint32_t num_cores = num_cores_c * num_cores_r;
     CoreRange all_device_cores(
-        {(std::size_t)start_core_x, (std::size_t)start_core_y},
-        {(std::size_t)start_core_x + num_cores_c - 1, (std::size_t)start_core_y + num_cores_r - 1});
+        {static_cast<std::size_t>(start_core_x), static_cast<std::size_t>(start_core_y)},
+        {static_cast<std::size_t>(start_core_x) + num_cores_c - 1,
+         static_cast<std::size_t>(start_core_y) + num_cores_r - 1});
```
[the remark this answers is not in the document]
Comment (dgomezTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_program_factory_attention_optimized_sharded.cpp]: Fixed in 144c407c8ec — `all_device_cores` is now declared as `const CoreRangeSet{CoreRange(...)}`.
